### PR TITLE
Fix zero dtcldp sensitivity without reintroducing dwdp blow-up

### LINKED ISSUE #3249: Forward sensitivities silently wrong (reported as 0) when a boundary-condition species' initial value depends on a parameter
### Summary

For a species with `boundaryCondition="true"` whose initial value is set via
`InitialAssignment` from a free parameter, AMICI correctly computes the
sensitivity of the species' own (constant) value, but fails to propagate
that sensitivity into other species whose reaction rates depend on it. The
result is a forward-sensitivity (`sx`) that is silently reported as exactly
zero when it should be nonzero and growing over time.

Found via finite-difference sensitivity checks newly added to AMICI's SBML
semantic test suite (`tests/sbml/testSBMLSuite.py`); confirmed against raw
`rdata.sx` independent of that check, and independent of PEtab/fiddy
entirely. Affects at least 12 cases in the official SBML semantic test
suite (case IDs 00498–00509).

### Reproducer

Two files, no dependencies beyond AMICI itself (`amici==1.1.0`, commit
`9456396bf`, plain SBML import — no PEtab).

`model.xml`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<sbml xmlns="http://www.sbml.org/sbml/level3/version2/core" level="3" version="2">
  <model id="boundary_species_sensitivity_bug" timeUnits="second">
    <listOfCompartments>
      <compartment id="C" spatialDimensions="3" size="1" constant="true"/>
    </listOfCompartments>
    <listOfSpecies>
      <species id="S1" compartment="C" boundaryCondition="true" hasOnlySubstanceUnits="true" constant="false"/>
      <species id="S2" compartment="C" initialAmount="1.5" hasOnlySubstanceUnits="true" boundaryCondition="false" constant="false"/>
    </listOfSpecies>
    <listOfParameters>
      <parameter id="k1" value="0.75" constant="true"/>
      <parameter id="k2" value="50" constant="true"/>
    </listOfParameters>
    <listOfInitialAssignments>
      <initialAssignment symbol="S1">
        <math xmlns="http://www.w3.org/1998/Math/MathML">
          <apply><times/><ci>k1</ci><ci>S2</ci></apply>
        </math>
      </initialAssignment>
    </listOfInitialAssignments>
    <listOfReactions>
      <reaction id="R1" reversible="false">
        <listOfReactants>
          <speciesReference species="S1" constant="true"/>
        </listOfReactants>
        <listOfProducts>
          <speciesReference species="S2" constant="true"/>
        </listOfProducts>
        <kineticLaw>
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply><times/><ci>C</ci><ci>k2</ci><ci>S1</ci></apply>
          </math>
        </kineticLaw>
      </reaction>
    </listOfReactions>
  </model>
</sbml>
```

`repro.py`:
```python
import logging
logging.disable(logging.CRITICAL)

import amici
import numpy as np
from amici.sim.sundials import SensitivityOrder, run_simulation
from pathlib import Path

sbml_file = Path(__file__).parent / "model.xml"
importer = amici.SbmlImporter(str(sbml_file))
importer.sbml2amici(
    "BoundarySpeciesBug",
    output_dir=Path(__file__).parent / "_model",
    generate_sensitivity_code=True,
)
model_module = amici.import_model_module(
    "BoundarySpeciesBug", Path(__file__).parent / "_model"
)
model = model_module.get_model()
model.set_timepoints(np.linspace(0, 4, 5))

solver = model.create_solver()
solver.set_sensitivity_order(SensitivityOrder.first)
rdata = run_simulation(model=model, solver=solver)

print("free parameters:", model.get_free_parameter_ids())
print("x0 (S1, S2):     ", rdata.x0)
print("sx0 (dS1/dk1, dS2/dk1 at t=0):", rdata.sx0[:, 0])
print()
print("x(t)  [S1, S2] over time:")
print(rdata.x)
print()
print("sx(t)[:, :, 0]  (d/dk1 of [S1, S2] over time -- S2 column should grow, not stay 0):")
print(rdata.sx[:, :, 0])


# finite-difference cross-check, independent of AMICI's sx entirely
def simulate_S2(k1):
    model.set_free_parameter_by_id({"k1": k1, "k2": 50.0})
    solver_fd = model.create_solver()
    solver_fd.set_sensitivity_order(SensitivityOrder.none)
    return run_simulation(model=model, solver=solver_fd).x[:, 1]


h = 1e-6
fd = (simulate_S2(0.75 + h) - simulate_S2(0.75 - h)) / (2 * h)
print()
print("finite-difference dS2(t)/dk1:", fd)
print("AMICI-reported    dS2(t)/dk1:", rdata.sx[:, 1, 0])
```

### Output

```
free parameters: ('k1', 'k2')
x0 (S1, S2):      [1.125 1.5  ]
sx0 (dS1/dk1, dS2/dk1 at t=0): [1.5 0. ]

x(t)  [S1, S2] over time:
[[  1.125   1.5  ]
 [  1.125  57.75 ]
 [  1.125 114.   ]
 [  1.125 170.25 ]
 [  1.125 226.5  ]]

sx(t)[:, :, 0]  (d/dk1 of [S1, S2] over time -- S2 column should grow, not stay 0):
[[1.5 0. ]
 [1.5 0. ]
 [1.5 0. ]
 [1.5 0. ]
 [1.5 0. ]]

finite-difference dS2(t)/dk1: [  0.          75.         150.         225.         300.00000001]
AMICI-reported    dS2(t)/dk1: [0. 0. 0. 0. 0.]
```

### Analysis

- `sx0[S1, k1] = 1.5` is correct: `S1(0) = k1 * S2(0) = 0.75 * 1.5 = 1.125`,
  so `dS1(0)/dk1 = S2(0) = 1.5`.
- `S1` is a boundary-condition species, so `xdot[S1] = 0` identically and
  `sx[S1](t)` correctly stays constant at its initial value for all `t`.
- `S1`'s value still appears in `S2`'s reaction rate (`C * k2 * S1`), so the
  chain-rule term `∂f_S2/∂S1 · sx[S1]` should carry that nonzero, constant
  1.5 forward into a linearly growing `sx[S2](t)`. Finite differences
  confirm this: `75, 150, 225, 300` at `t = 1..4`, matching
  `C·k2·(dS1/dk1)·t = 1·50·1.5·t = 75t`.
- AMICI reports `sx[S2](t) = 0` for every `t > 0` instead.

This looks like a state whose own sensitivity ODE is trivial
(`d(sx)/dt = 0`, because `xdot = 0` identically for a boundary-condition
species) is being dropped from the Jacobian used to propagate *other*
states' sensitivities forward — even though the state's *value* still
enters other states' rate laws. Suspected location: wherever
`de_export.py` gives boundary-condition species special treatment when
building the symbolic state Jacobian.

### Affected test-suite cases

SBML semantic test suite case IDs 00498–00509 (a boundary species set via
`InitialAssignment` from a free parameter, feeding a reaction rate) all
show this same pattern when FD-checked.

### Environment

- AMICI 1.1.0, commit `9456396bf`
- Plain SBML import (`amici.SbmlImporter`), no PEtab involved

---
🤖 Found via an AI-assisted (Claude) audit of AMICI's SBML semantic test suite sensitivity checks.

### LINKED ISSUE #3268: Always-symbolic dtcldp (#3259) blows up generated code size, hangs compilation for large models
🤖 Drafted with AI assistance.

#3259 made `dtcldp` (sensitivity of a conservation law's total abundance w.r.t. free parameters) always a live symbol, never folded to a literal `0`, to fix incorrect zero sensitivities for single-species conservation laws with parameter-dependent initial values.

This is currently hanging/timing out compilation for large models. Confirmed on the `CS_Signalling_ERBB_RAS_AKT` performance-test benchmark:

- `main` before #3259: generated model code ~758 KB, compiles fine.
- `main` at #3259: generated model code ~15.5 MB. `dwdp.cpp` alone grows from 693 KB / 10,012 lines to 148 MB / 2,228,620 lines. Compilation doesn't finish within the CI timeout (observed >2h, still running after 5h in one merge-queue attempt).
- The merge-queue run for #3259 alone (no other PR involved) reproduces the hang, ruling out any other change as the cause.

**Root cause:** `_dtcldp_symbols()` (`de_model.py`, `DEModel._dtcldp_symbols`) builds `dtcldp` as a fully **dense** matrix: one fresh, opaque symbol per `(conservation law, parameter)` pair, regardless of whether that parameter could plausibly affect that law. Because these symbols are opaque, sympy can never simplify them to zero downstream, so every pair gets threaded through `w`/`dwdp`. For this model (~1600 single-species/boundary conservation laws × ~460 free parameters), that's ~740k forced-nonzero entries added to `dwdp`'s sparsity pattern (nnz: 4,216 → 743,752) — nearly all of them trivial 1:1 copies (`dSP_10_..._ = stcl_SP_10_..._;`), but the sheer count overwhelms the compiler.

This currently blocks the merge queue for any PR targeting `main` behind #3259 (Performance Test job hangs until CI timeout).

cc @dweindl

### PR DESCRIPTION
**Pull request by dweindl**
Fixes #3249, fixes #3268.

`dtcldp` was hardcoded to `0` for every parameter whenever a conservation law spanned a single species — wrong whenever that species' own initial value depends on a parameter (#3249). #3259 fixed this by making `dtcldp` always symbolic for every `(conservation law, parameter)` pair, which blew up `dwdp`'s sparsity pattern for models with many trivial single-species conservation laws, hanging compilation (#3268); it was reverted.

Single-species conservation laws always come from states with `dx/dt == 0` identically, so their `dtcldp` sensitivity is fully decidable from the state's own initial-value formula: this emits a symbol only for parameters that actually appear there, literal zero otherwise. Multi-species conservation laws stay fully symbolic, since preequilibration followed by partial reinitialization can make their total depend on parameters invisible to any static formula — this is what PEtab test suite cases 0010/0017 protect against.

🤖 Generated with AI assistance.

:eyes: https://github.com/dweindl/AMICI/actions/runs/34533796660

### CODE REVIEW

File: python/sdist/amici/_symbolic/de_model.py [hunk 1]
```diff
@@ -2474,6 +2440,47 @@ def conservation_law_has_multispecies(self, tcl: ConservationLaw) -> bool:
...
+        """
+        Builds the symbol matrix for ``dtcldp``, the sensitivity of each
+        conservation law's total abundance w.r.t. the free parameters.
+
+        Multi-species conservation laws keep every entry as a live symbol:
+        preequilibration can make their total depend on any parameter that
+        affects the dynamics, not just ones visible in the model's own
+        ``x0``/``sx0`` formulas. Single-species conservation laws only ever
+        arise from states with ``dx/dt == 0`` identically, which
+        preequilibration can't move -- so their sensitivity is fully
```
Comment (FFroehlich) [inline python/sdist/amici/_symbolic/de_model.py]: Even though this implementation looks correct, I just realised that I don’t really agree with the remit that preeq can introduce non-zero sensitivities for totals. Either those entities are conserved and the total does not change and preeq can’t introduce non-zero sensitivities (eqs probably simplify to zero in that case anyways?), or they are not conserved. Are we missing something else? Reinitialisation?

**Review by FFroehlich — APPROVED** [review]

File: python/sdist/amici/_symbolic/de_model.py (hunk 1 above)
Comment (dweindl) [reply python/sdist/amici/_symbolic/de_model.py]: Yeah, you are right, it's not preeq as such causing the problem, it's the following partial reinitialization. Will reword.
